Kuadrant/kuadrant-controller
# ratelimitpolicy virtualhosts ratelimits

RateLimitPolicy add rate limit list at VirtualHost and route level.
 * At VirtualHost level, an array of RateLimit object including `stage` and `action list`. Matches Envoy's [VirtualHost API](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route_components.proto.html?highlight=virtualhost)
  * At route level, an array of RateLimit object including `stage` and `action list`. Matches Envoy's [RouteAction API](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route_components.proto.html?highlight=virtualhost#envoy-v3-api-msg-config-route-v3-routeaction)

```yaml
---
apiVersion: apim.kuadrant.io/v1alpha1
kind: RateLimitPolicy
metadata:
  name: toystore
spec:
  networkingRef:
    - type: VirtualService
      name: toystore
  routes:
    - name: get-toy
      rateLimits:
        - stage: BOTH
          actions:
            - generic_key:
                descriptor_key: get-toy
                descriptor_value: "yes"
            - generic_key:
                descriptor_key: other-get-toy
                descriptor_value: "yes"
    - name: add-toy
      rateLimits:
        - stage: POSTAUTH
          actions:
            - generic_key:
                descriptor_key: add-toy
                descriptor_value: "yes"
    - name: delete-toy
      rateLimits:
        - stage: PREAUTH
          actions:
            - generic_key:
                descriptor_key: delete-toy
                descriptor_value: "yes"
  rateLimits:
    - stage: PREAUTH
      actions:
        - generic_key:
            descriptor_key: vhaction
            descriptor_value: "yes"

  limits:
    - conditions: ["get-toy == yes"]
      max_value: 2
      namespace: preauth
      seconds: 15
      variables: []
```
## Verification steps
1. `make local-setup`
2. `k apply -f examples/toystore`

The RateLimitController should create 2 envoy filters:
```
k get envoyfilters.networking.istio.io -A
NAMESPACE         NAME                                         AGE
kuadrant-system   kuadrant-gateway-ratelimit-filters           33m
kuadrant-system   ratelimits-of-toystore-in-kuadrant-gateway   28m
```
The RateLimitController should create 1 rate limit:
```
k get ratelimits -A
NAMESPACE   NAME               AGE
default     toystore-limit-1   46m
```
REading the envoy conf, the rate limits should be there at VH level and route level:
```
kubectl port-forward --namespace kuadrant-system deployment/kuadrant-gateway 15000:15000
```
```
curl -v "http://127.0.0.1:15000/config_dump?resource=dynamic_route_configs&format=json" | yq e -P -
```
If the stage is `BOTH`, the rate limit config should be duplicated, one of them with stage PRE `0` and the other with stage POST `1`
